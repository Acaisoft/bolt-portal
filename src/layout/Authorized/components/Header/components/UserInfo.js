import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import { withRouter } from 'react-router-dom'
import { IconButton, Menu, MenuItem, withStyles } from '@material-ui/core'
import { AccountCircle } from '@material-ui/icons'

import styles from './UserInfo.styles'

export class UserInfo extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
    logout: PropTypes.func.isRequired,
  }

  handlers = {}

  state = {
    anchor: null,
  }

  render() {
    const { classes } = this.props
    const { anchor } = this.state
    const isOpen = !!anchor

    return (
      <div className={classes.root}>
        <IconButton
          aria-owns={isOpen ? 'menu-appbar' : null}
          aria-haspopup="true"
          onClick={this.handleMenuOpen}
          color="inherit"
        >
          <AccountCircle />
        </IconButton>
        <Menu
          id="menu-appbar"
          anchorEl={anchor}
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'right',
          }}
          transformOrigin={{
            vertical: 'top',
            horizontal: 'right',
          }}
          open={isOpen}
          onClose={this.handleMenuClose}
        >
          <MenuItem onClick={this.handleMenuLink('/products')}>Products</MenuItem>
          <MenuItem onClick={this.handleMenuLink('/jobs')}>Jobs</MenuItem>
          <MenuItem onClick={this.handleMenuLink('/profiles')}>Profiles</MenuItem>
          <MenuItem onClick={this.handleLogout}>Log Out</MenuItem>
        </Menu>
      </div>
    )
  }

  handleMenuOpen = event => {
    this.setState({ anchor: event.currentTarget })
  }

  handleMenuClose = () => {
    this.setState({ anchor: null })
  }

  // Cache generated handlers to prevent redundant re-renders.
  handleMenuLink = route => {
    if (typeof this.handlers[route] !== 'function') {
      const { history } = this.props
      const handler = () => {
        history.push(route)
        this.handleMenuClose()
      }
      this.handlers[route] = handler
    }

    return this.handlers[route]
  }

  handleLogout = () => {
    this.props.logout()
  }
}

const mapDispatch = () => ({
  logout: () => {},
})

export default withRouter(
  connect(
    null,
    mapDispatch
  )(withStyles(styles)(UserInfo))
)
