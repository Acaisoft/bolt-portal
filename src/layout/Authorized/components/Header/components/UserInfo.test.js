import React from 'react'
import { shallow } from 'enzyme'

import { UserInfo } from './UserInfo'
import { mocks } from '~utils/tests'

const initUserInfo = overrides => {
  const mockProps = {
    history: {
      push: jest.fn(),
    },
    logout: jest.fn(),
  }
  const wrapper = shallow(
    <UserInfo classes={mocks.ClassesProxy} {...mockProps} {...overrides} />
  )
  const instance = wrapper.instance()

  return { instance, mockProps, wrapper }
}

describe('layout: UserInfo', () => {
  describe('rendering', () => {
    it('should render without crashing', () => {
      const { wrapper } = initUserInfo()
      expect(wrapper).toBeTruthy()
    })

    it('should render properly', () => {
      const { wrapper } = initUserInfo()
      expect(wrapper).toMatchSnapshot()
    })

    it('should render open menu properly', () => {
      const { wrapper } = initUserInfo()
      wrapper.setState({ anchor: 'fake ref' })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('events', () => {
    describe('handleMenuOpen', () => {
      it('should set menu anchor', () => {
        const fakeEvent = { currentTarget: 'fake current target' }
        const { instance } = initUserInfo()

        instance.handleMenuOpen(fakeEvent)

        expect(instance.state.anchor).toBe('fake current target')
      })
    })

    describe('handleMenuClose', () => {
      it('should clear menu anchor', () => {
        const { instance } = initUserInfo()

        instance.handleMenuClose()

        expect(instance.state.anchor).toBe(null)
      })
    })

    describe('handleLogout', () => {
      it('should call logout action', () => {
        const { instance, mockProps } = initUserInfo()

        instance.handleLogout()

        expect(mockProps.logout).toHaveBeenCalled()
      })
    })

    describe('handleMenuLink', () => {
      it('should return a function', () => {
        const { instance } = initUserInfo()

        const fn = instance.handleMenuLink('/example')

        expect(typeof fn).toBe('function')
      })

      it('should cache functions with the same route', () => {
        const { instance } = initUserInfo()

        const fn1 = instance.handleMenuLink('/example')
        const fn2 = instance.handleMenuLink('/example')

        expect(fn1).toBe(fn2)
      })

      it('should redirect to the route and close menu', () => {
        const { instance, mockProps } = initUserInfo()
        instance.handleMenuClose = jest.fn()

        const fn = instance.handleMenuLink('/example')
        fn()

        expect(mockProps.history.push).toHaveBeenCalledWith('/example')
        expect(instance.handleMenuClose).toHaveBeenCalled()
      })
    })
  })
})
