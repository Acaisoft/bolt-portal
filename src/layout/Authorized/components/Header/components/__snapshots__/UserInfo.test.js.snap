// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout: UserInfo rendering should render open menu properly 1`] = `
<div
  className="root"
>
  <IconButtonMock
    aria-haspopup="true"
    aria-owns="menu-appbar"
    color="inherit"
    onClick={[Function]}
  >
    <AccountCircleMock />
  </IconButtonMock>
  <MenuMock
    anchorEl="fake ref"
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    id="menu-appbar"
    onClose={[Function]}
    open={true}
    transformOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
  >
    <MenuItemMock
      onClick={[Function]}
    >
      Products
    </MenuItemMock>
    <MenuItemMock
      onClick={[Function]}
    >
      Jobs
    </MenuItemMock>
    <MenuItemMock
      onClick={[Function]}
    >
      Profiles
    </MenuItemMock>
    <MenuItemMock
      onClick={[Function]}
    >
      Log Out
    </MenuItemMock>
  </MenuMock>
</div>
`;

exports[`layout: UserInfo rendering should render properly 1`] = `
<div
  className="root"
>
  <IconButtonMock
    aria-haspopup="true"
    aria-owns={null}
    color="inherit"
    onClick={[Function]}
  >
    <AccountCircleMock />
  </IconButtonMock>
  <MenuMock
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    id="menu-appbar"
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
  >
    <MenuItemMock
      onClick={[Function]}
    >
      Products
    </MenuItemMock>
    <MenuItemMock
      onClick={[Function]}
    >
      Jobs
    </MenuItemMock>
    <MenuItemMock
      onClick={[Function]}
    >
      Profiles
    </MenuItemMock>
    <MenuItemMock
      onClick={[Function]}
    >
      Log Out
    </MenuItemMock>
  </MenuMock>
</div>
`;
